### PR TITLE
Grid search tuner

### PR DESCRIPTION
**Pull request by liuzhe-lz**
### Description ###

Rewrite grid search tuner to support nested search space and uniform/normal.

This PR should be merged after TPE.

### Checklist ###
  - [x] test case
  - [x] doc

### How to test ###

Create an experiment with tuner name 'gridsearch'.
[Here](https://github.com/microsoft/nni/blob/master/test/ut/sdk/test_hpo_formatting.py#L4) is a sophisticated search space suite for test.

### CODE REVIEW

File: test/ut/sdk/test_builtin_tuners.py [hunk 1]
```diff
@@ -253,7 +253,7 @@ def import_data_test_for_pbt(self):
         logger.info("Imported data successfully at the beginning with incomplete epoch")
         shutil.rmtree(all_checkpoint_dir)
 
-    def import_data_test(self, tuner_factory, stype="choice_str"):
+    def import_data_test(self, tuner_factory, stype="choice_str", support_middle=True):
```
Comment (ultmaster) [inline test/ut/sdk/test_builtin_tuners.py:256]: What is `support_middle`

File: nni/algorithms/hpo/gridsearch_tuner.py [hunk 2]
```diff
@@ -2,209 +2,241 @@
...
+    return real_x < real_y
+
+def _cdf_inverse(x, spec):
+    # inverse function of spec's cumulative distribution function
+    if spec.normal_distributed:
+        return spec.mu + spec.sigma * math.sqrt(2) * erfinv(2 * x - 1)
+    else:
+        return spec.low + (spec.high - spec.low) * x
+
+def _deformat_single_parameter(x, spec):
```
Comment (ultmaster) [inline nni/algorithms/hpo/gridsearch_tuner.py:227]: Why is this needed provided that we have `deformat_parameters` already?

Comment (liuzhe-lz) [reply nni/algorithms/hpo/gridsearch_tuner.py:227]: Because it's not that easy to get a parameter inside nested space.
It either needs a function to do `while ...: obj = getattr(obj, ...)`, or this.
I think the complexity is similar, and chose this approach because it has better efficiency.

**Review by ultmaster — APPROVED** [review]

File: nni/algorithms/hpo/gridsearch_tuner.py [hunk 3]
```diff
@@ -2,209 +2,241 @@
...
+#
+# We divide normal distributed spaces using inverse function of CDF.
+# For example the 1/4 point of a normal distribution is defined as X where `normal_cdf(X) = 1/4`.
+#
+# Here is an example:
+#
+#   search space:
+#     x: choices(5, 7)
+#     y: normal(0, 1)
+#     z: quniform(2, 3, 1)
```
Comment (QuanluZhang) [inline nni/algorithms/hpo/gridsearch_tuner.py:47]: quniform is discrete space, you treat it a continuous space?

File: nni/algorithms/hpo/gridsearch_tuner.py [hunk 4]
```diff
@@ -2,209 +2,241 @@
...
+            else:
+                params[spec.key] = _cdf_inverse(x, spec)
+        return params
+
+def _less(x, y, spec):
+    if spec.q is None:
+        return x < y
+    real_x = _deformat_single_parameter(_cdf_inverse(x, spec), spec)
+    real_y = _deformat_single_parameter(_cdf_inverse(y, spec), spec)
+    return real_x < real_y
```
Comment (QuanluZhang) [inline nni/algorithms/hpo/gridsearch_tuner.py:218]: in which case, x < y but real_x > real_y?

File: nni/algorithms/hpo/gridsearch_tuner.py [hunk 5]
```diff
@@ -2,209 +2,241 @@
...
+    def receive_trial_result(self, *args, **kwargs):
+        pass
 
-    Type ``choice`` will select one of the options. Note that it can also be nested.
+    def import_data(self, data):  # for resuming
+        # this is not accurate for concurrency > 1
+        # but I believe it's not worthy to introduce complex logic to fix this
+        # instead we should support checkpoint (TODO)
+        for _ in data:
+            self._suggest()
```
Comment (QuanluZhang) [inline nni/algorithms/hpo/gridsearch_tuner.py:111]: this is not a correct logic for import data. import_data is not only for resume

File: nni/algorithms/hpo/gridsearch_tuner.py [hunk 6]
```diff
@@ -2,209 +2,241 @@
...
-                        else:
-                            chosen_params.append(choice)
-                elif _type == 'quniform':
-                    chosen_params = self._parse_quniform(_value)
-                elif _type == 'randint':
-                    chosen_params = self._parse_randint(_value)
-                else:
-                    raise RuntimeError("Not supported type: %s" % _type)
+            if self.vector is None:  # epoch end, update grid and retry
+                self._next_grid()
```
Comment (QuanluZhang) [inline nni/algorithms/hpo/gridsearch_tuner.py:122]: if there is a uniform, this grid search will never end, right?

File: nni/algorithms/hpo/gridsearch_tuner.py (hunk 3 above)
Comment (cruiseliu) [reply nni/algorithms/hpo/gridsearch_tuner.py:47]: The example has already make it clear.
`3/4 is eliminated due to duplication`

File: nni/algorithms/hpo/gridsearch_tuner.py (hunk 4 above)
Comment (cruiseliu) [reply nni/algorithms/hpo/gridsearch_tuner.py:218]: With "q", real_x can equal to real_y.

File: nni/algorithms/hpo/gridsearch_tuner.py (hunk 5 above)
Comment (cruiseliu) [reply nni/algorithms/hpo/gridsearch_tuner.py:111]: There is no way to invoke `import_data` except resuming.

File: nni/algorithms/hpo/gridsearch_tuner.py (hunk 6 above)
Comment (cruiseliu) [reply nni/algorithms/hpo/gridsearch_tuner.py:122]: Yes.

File: nni/algorithms/hpo/gridsearch_tuner.py (hunk 5 above)
Comment (QuanluZhang) [reply nni/algorithms/hpo/gridsearch_tuner.py:111]: we provided import data in nnictl, you removed it from nnictl?

Comment (cruiseliu) [reply nni/algorithms/hpo/gridsearch_tuner.py:111]: OK my bad. I have never noticed this command.
But I insist it should be fixed later in another PR.

Comment (QuanluZhang) [reply nni/algorithms/hpo/gridsearch_tuner.py:111]: we need to know that this is a broken change

Comment (cruiseliu) [reply nni/algorithms/hpo/gridsearch_tuner.py:111]: It can be fixed in this release. It won't take much time.
I don't want to fix it here because it feels like a new work item. Development of this PR and Yuge's review was finished a week ago so significant logic change should trigger a new round of PR and review.

And, it sounds silly to "import data" to grid search. Don't think [anyone will try it](https://github.com/microsoft/nni/issues?q=is%3Aissue+%22nnictl+experiment+import%22).

**Review by QuanluZhang — APPROVED** [review]
